### ISSUE TITLE
Basis: Needs styling on Node links like "Read more" and "Comments" 

### ISSUE THREAD

**Issue by quicksketch** [backdrop/backdrop-issues#2797]
## Describe your issue or idea
The styling on the node links could use some work in Basis. Right now they stack on top of each other, rather than displaying one after the other:

![image](https://user-images.githubusercontent.com/100206/29429319-25be0240-8345-11e7-96b8-4edaee6ab286.png)

### Steps to reproduce (if reporting a bug)

- Install Backdrop.
- Enable comments on the first post.
- Add comments to the first post.
- View the homepage.

**Comment by BWPanda** [comment]
Confirmed; still an issue in the latest version:

![IMG_20220605_120240](https://user-images.githubusercontent.com/2385329/172031787-f4912c08-f563-4d96-875f-1281dd8acbe1.jpg)

**Comment by BWPanda** [comment]
The PR was simple enough: https://github.com/backdrop/backdrop/pull/4122

The implications of it, on the other hand, might be trickier to address...

**Comment by BWPanda** [comment]
![image](https://user-images.githubusercontent.com/2385329/177036422-d819b158-1d39-4333-8f08-b932d5250505.png)

**Comment by argiepiano** [comment]
I find the comma separator in "Read more, 1 comment, Add comment" rather confusing. The commas makes them seem more semantically related than they are (not to mention that "Add" should not be capitalized after a comma, but this is not this PR's fault).  Specifically "Read more, 1 comment" makes it seem like the Read more is somehow related to comments, but it's not.

**Comment by bugfolder** [comment]
For parallelness of construction, grammar, and punctuation, should it be

[Read more](#); [see 1 comment](#); [add comment](#)

?

**Comment by bugfolder** [comment]
I added another PR that fixes the capitalization of "Add" in addition to the CSS change of the existing PR.

![image](https://github.com/backdrop/backdrop-issues/assets/29214084/82eae1d2-9985-40a7-8927-a1822bf48c7d)

There's still the question of whether inline links should be delimited with something "stronger" than commas (like semicolons, in my comment just above).

Even though semicolons are more grammatical in this example, since inline links are used in a lot of places, it's not clear that they should be replaced everywhere. If it's a good idea for here, then maybe should we support another CSS class that replaces the commas with semicolons in specific locations?

**Comment by bugfolder** [comment]
Meanwhile, tests need to be updated for this PR (to check for consistent capitalization).

**Comment by olafgrabienski** [comment]
> I added another PR that fixes the capitalization of "Add" in addition to the CSS change of the existing PR.

The fix works for a post with comments (screenshot No. 1), but not for content *without* comments (No. 2).

**No. 1, Post with comments**
<img src="https://github.com/backdrop/backdrop-issues/assets/14188307/85bd81b6-6e79-4f3e-bf83-e6043dabbb70" width="280">

**No. 2, without comments**
<img src="https://github.com/backdrop/backdrop-issues/assets/14188307/3d187242-a74f-4118-8de0-d2e5f074ebb7" width="280">

**Comment by olafgrabienski** [comment]
> There's still the question of whether inline links should be delimited with something "stronger" than commas (like semicolons, in my comment just above).

I agree, commas aren't the best option here, but I don't like semicolons either. In my opinion, both are misleading, suggesting that inline links are part of a grammatical sentence. I see them rather as independent elements. To make this more clear, they could be separated by an element which is normally not used grammatically (not as punctuation in the strict sense), and to keep the "Add" capitalized. Ideas:

Slash:
`Read more / 1 comment / Add comment`

Pipe:
`Read more | 1 comment | Add comment`

Middot:
`Read more · 1 comment · Add comment`

Spacing (CSS margin):
`Read more   1 comment   Add comment`

My favorites are the middot and spacing.

**Comment by olafgrabienski** [comment]
> *Read more, 1 comment, Add comment* (...) could be a list, if it listed actions to do on comments.

Interesting! I guess, that was the reason for @bugfolder's addition of a verb ("see comment") here: https://github.com/backdrop/backdrop-issues/issues/2797#issuecomment-1173103844

**Comment by bugfolder** [comment]
> Middot:
> `Read more · 1 comment · Add comment`

Middot is my favorite, too, and a nice thing about not using sentence punctuation is then capitalizing each element (current behavior) makes grammatical sense.

> that was the reason for @bugfolder's addition of a verb ("see comment") here

Yes, that way they're all imperative sentences.

**Comment by bugfolder** [comment]
I modified the PR to go back to always capitalizing and use heavy dots for delimiters (which seemed better balanced to me):

![image](https://github.com/backdrop/backdrop-issues/assets/29214084/76f70203-ddfa-49b8-a429-aea5d84b0ba9)

Using diferent delimiters for inline links seems to makes sense here, but what about generally?

I did a search in core for `'links', 'inline'` (to find instances of `'#attributes' => array('class' => array('links', 'inline'))`, and see examples in:

* book.module
* comment.entity.inc
* comment.module
* file.entity.inc
* node.entity.inc
* redirect_handler-field_redirect_operations.inc

As for changing `N comments` to `See N comments` for links, I'd like to hear more of others' thoughts on that.

**Comment by olafgrabienski** [comment]
> Using diferent delimiters for inline links seems to makes sense here, but what about generally?

I'd say it makes generally sense. It would be good to check with some examples, but I don't have the time at the moment.

Re changing `N comments` to `See N comments`, I prefer the first (less noisy). Actually, I think it deserves a separate issue.

**Comment by findlabnet** [comment]
I think things like different delimiters are a matter of user taste or styling rules, so it should be configurable just like breadcrumbs delimiters in some themes.

**Comment by bugfolder** [comment]
Here's an example using `redirect_handler_field_redirect_operations.inc`: create a View table of URL redirects, then add the Operations field. In Seven (in the View editor), it looks like this:

![image](https://github.com/backdrop/backdrop-issues/assets/29214084/83a251a8-dfd9-496d-aeb2-8dfb0ef05c70)

And in Basis, with the dot delimiters, it looks like this:

![image](https://github.com/backdrop/backdrop-issues/assets/29214084/a8cd16d9-88b4-4e8f-b31e-460f8dee4f7d)

I don't think it needs the dots here, but they don't hurt. (Actually, I'd prefer that the "Operations" render as dropbuttons, as is done in so many other places for multiple operations, but that would be a different issue.)

**Comment by bugfolder** [comment]
For the usage in `node.entity`, this theming is used for the "Read more" link in teasers, but since there's only one link, there's no delimiter. But modules could (via `hook_node_view` or `hook_entity_view` add additional links.)

**Comment by bugfolder** [comment]
For the usage in `file.entity`, the links are themed inline, but the list of links is empty unless augmented via `hook_file_view` or `hook_entity_view`.

**Comment by olafgrabienski** [comment]
@bugfolder Thanks for the examples! Looks like your statement about one of them (dots not needed, but they don't hurt) is valid for most places. However, with different ideas like dropbuttons under consideration, I'm not sure if it's worth the effort to change all the places now.

**Comment by olafgrabienski** [comment]
Sounds goood! I guess, #4167 comes into play now. Time to agree on a solution there!

**Comment by klonos** [comment]
I was under the impression that we were using pipes as link delimiters, but maybe I am remembering a specific theme, or perhaps somethings specific to GovCMS 🤔

**Comment by klonos** [comment]
...it was the Views UI!:
![image](https://github.com/backdrop/backdrop-issues/assets/2423362/d2d1f292-b24a-45b4-a416-b17918f886fb)

So you can say that that's an existing pattern we should retain I think.

**Comment by bugfolder** [comment]
> So you can say that that's an existing pattern we should retain I think.

I agree. So, to recap...before:

![](https://user-images.githubusercontent.com/100206/29429319-25be0240-8345-11e7-96b8-4edaee6ab286.png)

Updated PR:

![image](https://github.com/backdrop/backdrop-issues/assets/29214084/aee31025-8af3-4d67-9388-42be11d9b085)

**Comment by olafgrabienski** [comment]
Hm, looks I was a bit slow and the PR is already updated: Personally I don't like the pipes, too obtrusive.

Re consistency, the Views UI is a different thing, and it wasn't revised for a long time. We _can_ follow its pattern, but in my opinion we _don't have to_.

**Comment by bugfolder** [comment]
I guess I don't have a strong opinion myself (though consistency, however tenuous, won the day for me here). Since we clearly have a difference of opinion, how shall we settle this? (Perhaps our designer, @dariusgarza, could chime in?)

**Comment by olafgrabienski** [comment]
Thanks @bugfolder, and thanks for all the work on this. In a way, I have a strong opinion, but it's also a personal preference, and I don't want to derail the issue. The pipe symbol is also fine for me.

**Comment by kiamlaluno** [comment]
I find the pipes in the last screenshot fine too. I guess it may be because they seem like the borders of table cells.  
In the case the part after the delimiter could be rendered in the next line, I find the pipes more appropriate.

English does not use middle dots as punctuation anymore, but some languages use them. This could be a reason for avoiding middle dots to separate links.

**Comment by klonos** [comment]
Speaking of existing patterns, we have used middots to indicate indentation/nesting in issues like #4770, #4218 etc.

**Comment by olafgrabienski** [comment]
> Speaking of existing patterns, we have used middots to indicate indentation/nesting ...

Funny! When we introduced middots in #4218, it was a good choice because they didn't follow existing patterns ;-)

**Comment by bugfolder** [comment]
Seems like enough time and feedback has elapsed to to remove the "needs feedback." Looks like everyone thus far either prefers pipes or is at least OK with pipes, so I'll suggest we move on to the next phase of testing & code review. (Or express newly strong opposition, as the case may be.)

**Comment by stpaultim** [comment]
I'm good with the middots or pipes. My concern is whether or not any changes to this might break websites that have already sub-themed basis to fix this problem in their own manner?

**Comment by stpaultim** [comment]
I brought this issue to the developer meeting today to ask if we thought that this was backward compatible. The general consensus was, that based upon other decisions we've been making, this would not be considered backward compatible at this time. 

So, I think the official status of this issue would be on hold until we come up with a way to make backward compatible css changes. See: Supplemental CSS Selectors (or one of the other ideas out there). 

https://github.com/backdrop/backdrop-issues/issues/4782

**Comment by stpaultim** [comment]
We now have the ability to fix this in Basis. 

@bugfolder are you able to rebase the PR? We have until Jan 15th to get this into the next release.

**Comment by bugfolder** [comment]
PR is rebased (new PR), and RTL support is added.

LTR:
![image](https://github.com/user-attachments/assets/4391c7ac-5cdf-4a3e-b4e0-01a8cbc917d1)


RTL:
![image](https://github.com/user-attachments/assets/5c63f4b0-207d-45f7-84cc-9c86228f61c6)

**Comment by laryn** [comment]
I posted [a comment in Zulip](https://backdrop.zulipchat.com/#narrow/channel/218635-Backdrop/topic/Logical.20properties) asking about whether we can/should use logical properties instead of hard-coding RTL separately, going forward.

**Comment by bugfolder** [comment]
Here's the corresponding logical properties issue for Backdrop core: https://github.com/backdrop/backdrop-issues/issues/6657

Based on the Zulip discussion, making the switch to logical properties in core may take a while; I suggest that it need not hold up current PRs that contain LTR/RTL CSS (such as this one), so I'll suggest/request continued review/testing of the current PR.

**Comment by laryn** [comment]
Agreed.

**Comment by stpaultim** [comment]
It seems that this issue was blocked by the backwards compatibility issue. So, now it needs the `update-1-30` css class to ensure backward compatibility.

**Comment by bugfolder** [comment]
> now it needs the `update-1-30` css class to ensure backward compatibility.

@stpaultim, can you explain further, or link to where in the docs this class is explained? (I'm not seeing it on the docs site, except in test code.)

**Comment by bugfolder** [comment]
`update-1-30` class added to CSS selectors. Results:

* No update class on body, LTR:
![image](https://github.com/user-attachments/assets/c24b2b81-6eb7-4628-88ee-d96c9557ca0c)

* No update class on body, RTL:
![image](https://github.com/user-attachments/assets/888de318-17ce-4332-80df-8cd329e89ea5)

* Add `update-1-30` class to body, LTR:
![image](https://github.com/user-attachments/assets/0c6c2151-1fb2-48ba-8ea1-0d3c413b5139)

* Add `update-1-30` class to body, RTL:
![image](https://github.com/user-attachments/assets/1ed28177-f677-475e-8e20-7f797cfae465)

Ready again for review.

**Comment by stpaultim** [comment]
@bugfolder - You are correct, that we probably need some documentation for contributors on how to use the new CSS classes. I made a note of that here:

https://github.com/backdrop-ops/docs.backdropcms.org/issues/244

I'm waiting for feedback on the documentation that I already created.

Sorry, I had intended to provide some suggestions for you, but didn't get to it.

**Comment by stpaultim** [comment]
I like the pipes. Going to mark this WFM. It's been discussed a great deal already.
